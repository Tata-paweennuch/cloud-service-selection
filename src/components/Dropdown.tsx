import React from 'react';
import { Option } from './SharedInterface'

interface DropdownProps {
  label: string,
  id: string,
  value: string,
  handleOnChange: (name: string) => void,
  options: Option[]
}

const Dropdown = (props: DropdownProps) => {
  const {label, id, value, handleOnChange, options} = props

  const handleValueChange = (e: React.SyntheticEvent) => {
    const target  = e.target as HTMLSelectElement
    handleOnChange(target.value)
  }

  // try to use semantic elements, so we can get the best accessibility
  return (
    <div className="dropdown">
      <label htmlFor={id}>{ label }</label>
      <select name={id} id={id} onChange={handleValueChange} value={value}>
        {options.map((option, index) =>
          <option value={option.value} key={option.value + index}>{option.displayName}</option>
        )}
      </select>
    </div>  
  )  
}

export default Dropdown;
