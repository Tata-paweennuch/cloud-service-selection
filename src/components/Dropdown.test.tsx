import { shallow } from 'enzyme';
import Dropdown from './Dropdown';
import renderer from 'react-test-renderer'

let wrapper: any
let action: any
let options: any

beforeEach(() => {
  action = jest.fn()
  options = [
    { value: 'may', displayName: 'May' },
    { value: 'june', displayName: 'June' }
  ]
  wrapper = shallow(<Dropdown label="Please select" id="month" value="April" handleOnChange={action} options={options}  />)
})

afterEach(() => {
  if(wrapper) {
    wrapper.unmount()
    wrapper = null
  }
  action.mockClear()
})

test('should render correctly with options', async () => {
  const tree = renderer.create(<Dropdown label="Please select" id="month" value="April" handleOnChange={action} options={options}/>)
  expect(tree.toJSON()).toMatchSnapshot()
})

test('should display select and option elements with value correctly', () => {
  expect(wrapper.find('.dropdown').exists()).toBeTruthy
  expect(wrapper.find('label').text()).toEqual('Please select')
  expect(wrapper.find('label').props()['htmlFor']).toEqual('month')
  expect(wrapper.find('select').props()['name']).toEqual('month')
  expect(wrapper.find('select').props()['id']).toEqual('month')
  expect(wrapper.find('select').props()['value']).toEqual('April')
  // has 2 options
  expect(wrapper.find('option').length).toEqual(2)
  expect(wrapper.find('option').at(0).text()).toEqual(options[0].displayName)
  expect(wrapper.find('option').at(0).props()['value']).toEqual(options[0].value)
  expect(wrapper.find('option').at(1).text()).toEqual(options[1].displayName)
  expect(wrapper.find('option').at(1).props()['value']).toEqual(options[1].value)
})

test('should call handleOnChange props with correct argument when selecting an option', () => {
  wrapper.find('select').simulate('change', {
    target: { value: 'april' }
  })

  expect(action).toHaveBeenCalledTimes(1)
  expect(action).toHaveBeenLastCalledWith('april')
})
