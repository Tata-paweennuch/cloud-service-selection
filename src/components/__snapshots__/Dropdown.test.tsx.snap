// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with options 1`] = `
<div
  className="dropdown"
>
  <label
    htmlFor="month"
  >
    Please select
  </label>
  <select
    id="month"
    name="month"
    onChange={[Function]}
    value="April"
  >
    <option
      value="may"
    >
      May
    </option>
    <option
      value="june"
    >
      June
    </option>
  </select>
</div>
`;
